### PR TITLE
451006 - Ivan Shimko

### PR DESCRIPTION
**Pull request by vanashimko**

### CODE REVIEW

File: DTOFromJsonGenerator/DTOFromJsonGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using DTOGeneratorLibrary;
+using Newtonsoft.Json;
+using static DTOFromJsonGenerator.Properties.Settings;
+
+namespace DTOFromJsonGenerator
+{
+    [SuppressMessage("ReSharper", "InconsistentNaming")]
+    class DTOFromJsonGenerator
```
Comment (RamanBut-Husaim) [inline DTOFromJsonGenerator/DTOFromJsonGenerator.cs]: @vanashimko You could fix naming to be something like `DtoFromJsonGenerator`.

File: DTOFromJsonGenerator/DTOFromJsonGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            }
+            else
+            {
+                PrintErrorMessage("Invalid number of arguments");
+            }
+        }
+
+        // Static internals
+
+        private static void WriteResults(IEnumerable<DTOClassDeclaration> dtoClassesDeclaration, string outputDirectoryPath)
```
Comment (RamanBut-Husaim) [inline DTOFromJsonGenerator/DTOFromJsonGenerator.cs]: @vanashimko Please, move this stuff to a separate class that will be responsible for output creation.

File: DTOFromJsonGenerator/JsonToDTOInfo.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+using DTOGeneratorLibrary;
+
+namespace DTOFromJsonGenerator
+{
+    internal class JsonClassDescription
+    {
+        public string ClassName { get; set; }
+        public JsonPropertyDescription[] Properties { get; set; }
+
+        internal DTOClassInfo ToDTOClassInfo()
```
Comment (RamanBut-Husaim) [inline DTOFromJsonGenerator/JsonToDTOInfo.cs]: @vanashimko Method naming. The general rule - replace your `DTO` abbreviation with `Dto`.

File: DTOGeneratorLibrary/SupportedTypesTable.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public enum TypeKind
+    {
+        Integer,
+        Number,
+        Boolean,
+        String,
+        Object
+    }
+
+    public class TypeNotRegisteredException : Exception
```
Comment (RamanBut-Husaim) [inline DTOGeneratorLibrary/SupportedTypesTable.cs]: @vanashimko Mmm, take a look at [this](http://dailydotnettips.com/2015/01/01/simplest-way-of-creating-custom-exceptions-in-net-using-visual-studio-2015/) article on how to declare custom exceptions.

File: DTOGeneratorLibrary/TypeDescription.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+
+namespace DTOGeneratorLibrary
+{
+    public struct TypeDescription
+    {
+        public readonly TypeKind TypeKind;
```
Comment (RamanBut-Husaim) [inline DTOGeneratorLibrary/TypeDescription.cs]: @vanashimko Please, replace fields with properties.

File: DTOGeneratorLibrary/TypeDescription.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+
+namespace DTOGeneratorLibrary
+{
+    public struct TypeDescription
```
Comment (RamanBut-Husaim) [inline DTOGeneratorLibrary/TypeDescription.cs]: @vanashimko Why have you made `TypeDescription` a struct?

File: DTOGeneratorLibrary/SupportedTypesTable.cs (hunk 4 above)
Comment (vanashimko) [inline DTOGeneratorLibrary/SupportedTypesTable.cs]: Should I declare all Exception's constructors in my exceptions?

File: DTOFromJsonGenerator/JsonToDTOInfo.cs (hunk 3 above)
Comment (vanashimko) [inline DTOFromJsonGenerator/JsonToDTOInfo.cs]: What about namespaces with `DTO` abbreviation?

File: DTOGeneratorLibrary/SupportedTypesTable.cs (hunk 4 above)
Comment (RamanBut-Husaim) [inline DTOGeneratorLibrary/SupportedTypesTable.cs]: @vanashimko Only 4 constructors from the article.

File: DTOFromJsonGenerator/JsonToDTOInfo.cs (hunk 3 above)
Comment (RamanBut-Husaim) [inline DTOFromJsonGenerator/JsonToDTOInfo.cs]: @vanashimko Yep.

Comment (vanashimko) [inline DTOFromJsonGenerator/JsonToDTOInfo.cs]: And solution's name too?

File: DTOGeneratorLibrary/DTOGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+namespace DTOGeneratorLibrary
+{
+    [SuppressMessage("ReSharper", "InconsistentNaming")]
+    public class DTOGenerator
+    {
+        private int _runningTasksCount;        
+        private readonly Queue<TaskInfo> _tasksQueue = new Queue<TaskInfo>();
+        private readonly object _syncRoot = new object();
+        private CountdownEvent _countdownEvent;
```
Comment (RamanBut-Husaim) [inline DTOGeneratorLibrary/DTOGenerator.cs]: @vanashimko Wrapped disposable object. Take a look at [this](https://msdn.microsoft.com/en-us/library/b1yfkh5e(v=vs.110).aspx) article.

File: DTOGeneratorLibrary/DTOGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+        public DtoGenerator(int maxRunningTasksCount, string namespaceName)
+        {
+            NamespaceName = namespaceName;
+            MaxRunningTasksCount = maxRunningTasksCount;
+        }
+
+        public DtoClassDeclaration[] GenerateDtoClasses(DtoClassInfo[] dtoClassesInfo)
+        {
+            using (_countdownEvent = new CountdownEvent(dtoClassesInfo.Length))
```
Comment (RamanBut-Husaim) [inline DTOGeneratorLibrary/DTOGenerator.cs]: @vanashimko A very dangerous thing happens there actually. You recreate an event for each call of `GenerateDtoClasses`. The operations that you perform are time-consuming so you cannot say when they will be finished. At the same time you are using the event in other methods in your class. As a result you could get a strange behaviour if the user will call your method side by side.

**Comment by vanashimko** [comment]
@RamanBut-Husaim Please, take a look

**Pull request merged**